File: IvanShurenkov/Second Term/Block1/Task1/src/main/kotlin/filters/MedianFilter.kt
package filters

import BMPFile

class MedianFilter : Filter() {
    fun applyFilter(bmpFile: BMPFile) {
        val step: Int = (3 + (bmpFile.bitCnt == 32).toInt())
        for (clrNum in 0..2) {
            for (i in 1 until bmpFile.height - 1) {
                for (j in step + clrNum until bmpFile.row - step step step) {
                    var index: Int = 0
                    var arr: Array<Short> = Array(9) { 255 }
                    for (q in i - 1..i + 1) {
                        for (w in j - step..j + step step step) {
                            if (0 <= w && w < bmpFile.row && 0 <= q && q < bmpFile.height) {
                                arr[index] = bmpFile.image[q][w]
                                index++
                            }
                        }
                    }
                    arr.sort()
                    bmpFile.image[i][j] = arr[index / 2]
                }
            }
        }
    }
}

File: Oleg-Tsvetkov/Term 2/Block 1/BMPFilter/BMPFilter/Filters/MedianFilter.cs
namespace BMPFilter.Filters
{
    public static class MedianFilter
    {
        public static void ApplyFilter(BitMapFile image)
        {
            for (int color = 0; color < 3; ++color)
            {
                for (int i = 1; i < image.Height - 1; ++i)
                {
                    for (int j = image.Channels; j < image.RowByteSize - image.Channels; j += image.Channels)
                    {
                        int[] nearColors = new int[9];
                        for (int n = -1; n < 2; ++n)
                        {
                            for (int m = -1; m < 2; ++m)
                            {
                                nearColors[(n + 1) * 3 + m + 1] = image.ImageBytes[i + n][j + color + image.Channels * m];
                            }
                        }
                        Array.Sort(nearColors);
                        image.ImageBytes[i][j + color] = (byte)nearColors[4];
                    }
                }
            }
        }
    }
}